crates/storage/provider/src/providers/database/provider.rs

```diff
@@ -736,20 +735,66 @@ impl<'this, TX: DbTxMut<'this> + DbTx<'this>> DatabaseProvider<'this, TX> {
         Ok(())
     }
 
-    /// Load last shard and check if it is full and remove if it is not. If list is empty, last
+    /// Load shard and check if it is full and remove if it is not. If list is empty, last
     /// shard was full or there is no shards at all.
-    pub fn take_last_storage_shard(&self, address: Address, storage_key: H256) -> Result<Vec<u64>> {
-        let mut cursor = self.tx.cursor_read::<tables::StorageHistory>()?;
-        let last = cursor.seek_exact(StorageShardedKey::new(address, storage_key, u64::MAX))?;
-        if let Some((storage_shard_key, list)) = last {
+    fn take_shard<T>(&self, key: T::Key) -> Result<Vec<u64>>
+    where
+        T: Table<Value = BlockNumberList>,
+    {
+        let mut cursor = self.tx.cursor_read::<T>()?;
+        let last = cursor.seek_exact(key)?;
+        if let Some((shard_key, list)) = last {
             // delete old shard so new one can be inserted.
-            self.tx.delete::<tables::StorageHistory>(storage_shard_key, None)?;
+            self.tx.delete::<T>(shard_key, None)?;
             let list = list.iter(0).map(|i| i as u64).collect::<Vec<_>>();
             return Ok(list)
         }
         Ok(Vec::new())
     }
 
+    /// Insert history index to the database.
+    ///
+    /// For each updated partial key, this function removes the last shard from
+    /// the database (if any), appends the new indices to it, chunks the resulting integer list and
+    /// inserts the new shards back into the database.
+    ///
+    /// This function is used by history indexing stages.
+    fn insert_history_index<P, T>(
```

Comment:
IIUC this function is expected to be called in append way, so we can't insert indices that are less than the last inserted index. So just a small nit on naming for clarity
```suggestion
    fn append_history_index<P, T>(
```